lib/HTTP/Request.pm

```diff
@@ -92,7 +92,19 @@ sub uri
 sub uri_canonical
 {
     my $self = shift;
-    return $self->{'_uri_canonical'} ||= $self->{'_uri'}->canonical;
+
+    my $uri = $self->{_uri};
+
+    if (defined (my $canon = $self->{_uri_canonical})) {
+        # early bailout if these are the exact same string; try to use
+        # the cheapest comparison method possible
+        return $canon if $$canon eq $$uri;
+
+        # otherwise needs refreshing
+        undef $self->{_uri_canonical};
```

Comment:
Since it's being set unconditionally in the next statement, this line isn't really needed.

Reply:
ah yes you're right. update forthcoming.

Reply:
and done